flax/linen/recurrent.py

```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
     key1, key2 = random.split(rng)
     mem_shape = batch_dims + size
     return init_fn(key1, mem_shape), init_fn(key2, mem_shape)
+
+class RNN(Module):
+  """The ``RNN`` module takes any :class:`RNNCellBase` instance and applies it over a sequence
+  using :func:`flax.linen.scan`.
+
+  Example::
+
+    >>> import jax.numpy as jnp
+    >>> import jax
+    >>> import flax.linen as nn
+    ...
+    >>> x = jnp.ones((10, 50, 32)) # (batch, time, features)
+    >>> lstm = nn.RNN(nn.LSTMCell(), cell_size=64)
+    >>> variables = lstm.init(jax.random.PRNGKey(0), x)
+    >>> y = lstm.apply(variables, x)
+    >>> y.shape # (batch, time, cell_size)
+    (10, 50, 64)
+
+  RNN accepts inputs with shape ``(*batch, time, *features)`` by default, if you set
+  ``time_major=True`` it will instead expect inputs with shape ``(time, *batch, *features)``::
+
+    >>> x = jnp.ones((50, 10, 32)) # (time, batch, features)
+    >>> lstm = nn.RNN(nn.LSTMCell(), cell_size=64, time_major=True)
+    >>> variables = lstm.init(jax.random.PRNGKey(0), x)
+    >>> y = lstm.apply(variables, x)
+    >>> y.shape # (time, batch, cell_size)
+    (50, 10, 64)
+
+  The output is an array of shape ``(*batch, time, *cell_size)`` by default (typically), however
+  if you set ``return_carry=True`` it will instead return a tuple of the final carry and the output::
+
+    >>> x = jnp.ones((10, 50, 32)) # (batch, time, features)
+    >>> lstm = nn.RNN(nn.LSTMCell(), cell_size=64, return_carry=True)
+    >>> variables = lstm.init(jax.random.PRNGKey(0), x)
+    >>> carry, y = lstm.apply(variables, x)
+    >>> jax.tree_map(jnp.shape, carry) # ((batch, cell_size), (batch, cell_size))
+    ((10, 64), (10, 64))
+    >>> y.shape # (batch, time, cell_size)
+    (10, 50, 64)
+
+  To support variable length sequences, you can pass a ``segmentation_mask`` which is an integer
+  array of shape ``(*batch, time)`` 1s indicate the elements are part of the sequence and 0s indicate
+  padding elements. Sequences must be padded to the right, i.e. all elements of a sequence must be
+  contiguous and padded elements must be to the right of the sequence. For example::
+
+  TODO: improve example
+
+    >>> # 3 sequences with max length 5
+    >>> segmentation_mask = jnp.array([
```

Comment:
I forgot the discussion around this, but why don't we just provide a list of sequence lengths? Isn't that more concise? So here we would have `seq_len = jnp.array([3, 2, 5])`.

Reply:
This format is compatible with sequence packing so we kept it in case we want that feature in the future. If we want to ditch sequence packing I'd be in favor of the change.

Reply:
Comment added.